File: cpp-lib-sync/examples/conformance_agent.cpp
// Conformance test agent for the cross-language interop suite (C++ sync).
// Same CLI as every other agent; a sink prints each received message on its own
// line and exits after --count messages.
#include "aiomsg/socket.hpp"

#include <chrono>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <optional>
#include <string>
#include <thread>

namespace {

std::string arg(int argc, char** argv, const std::string& key, const std::string& def) {
    std::string flag = "--" + key;
    for (int i = 1; i + 1 < argc; i++) {
        if (flag == argv[i]) {
            return argv[i + 1];
        }
    }
    return def;
}

aiomsg::Identity parse_identity(const std::string& hex) {
    aiomsg::Identity id{};
    for (size_t i = 0; i < id.size() && i * 2 + 1 < hex.size(); i++) {
        id[i] = static_cast<uint8_t>(std::strtol(hex.substr(i * 2, 2).c_str(), nullptr, 16));
    }
    return id;
}

}  // namespace

int main(int argc, char** argv) {
    using namespace aiomsg;

    std::string role = arg(argc, argv, "role", "connect");
    std::string host = arg(argc, argv, "host", "127.0.0.1");
    uint16_t port = static_cast<uint16_t>(std::atoi(arg(argc, argv, "port", "25000").c_str()));
    std::string send_mode = arg(argc, argv, "send-mode", "roundrobin");
    std::string behavior = arg(argc, argv, "behavior", "sink");
    int count = std::atoi(arg(argc, argv, "count", "10").c_str());
    std::string prefix = arg(argc, argv, "prefix", "m");
    std::string delivery = arg(argc, argv, "delivery", "at-most-once");
    std::string identity_hex = arg(argc, argv, "identity", "");
    double linger = std::atof(arg(argc, argv, "linger", "1.0").c_str());
    bool tls = arg(argc, argv, "tls", "false") == "true";
    std::string tls_cert = arg(argc, argv, "tls-cert", "");
    std::string tls_key = arg(argc, argv, "tls-key", "");
    std::string tls_ca = arg(argc, argv, "tls-ca", "");
    std::string server_name = arg(argc, argv, "tls-server-name", "");

    SendMode mode = send_mode == "publish" ? SendMode::Publish : SendMode::RoundRobin;
    Delivery del = delivery == "at-least-once" ? Delivery::AtLeastOnce : Delivery::AtMostOnce;
    std::optional<Identity> id;
    if (!identity_hex.empty()) {
        id = parse_identity(identity_hex);
    }

    Socket sock(mode, del, id);
    try {
        if (role == "bind") {
            if (tls) {
                sock.bind(host, port, ServerTls{tls_cert, tls_key});
            } else {
                sock.bind(host, port);
            }
        } else {
            if (tls) {
                sock.connect(host, port, ClientTls{tls_ca, server_name});
            } else {
                sock.connect(host, port);
            }
        }
    } catch (const std::exception& e) {
        std::fprintf(stderr, "%s failed: %s\n", role.c_str(), e.what());
        return 1;
    }

    auto linger_sleep = [&]() {
        std::this_thread::sleep_for(std::chrono::milliseconds(static_cast<int>(linger * 1000)));
    };

    if (behavior == "source") {
        for (int i = 0; i < count; i++) {
            std::string msg = prefix + std::to_string(i);
            sock.send(msg);
        }
        linger_sleep();
    } else if (behavior == "echo") {
        for (int i = 0; i < count; i++) {
            auto m = sock.recv();
            if (!m) {
                break;
            }
            sock.send_to(m->sender, m->data);
        }
        linger_sleep();
    } else {  // sink
        for (int i = 0; i < count; i++) {
            auto m = sock.recv();
            if (!m) {
                break;
            }
            std::fwrite(m->data.data(), 1, m->data.size(), stdout);
            std::fputc('\n', stdout);
            std::fflush(stdout);
        }
    }

    sock.close();
    return 0;
}

File: rust-lib-async/examples/conformance_agent.rs
//! Conformance test agent for the cross-language interop suite.
//!
//! A tiny, uniform CLI that the `conformance/` runner drives. The Python
//! implementation ships an agent with the same flags, so the runner can pair
//! any two languages in any role and assert they interoperate on the wire.
//!
//! Flags (all `--key value`):
//!   --role      bind | connect
//!   --host      e.g. 127.0.0.1
//!   --port      e.g. 25000
//!   --send-mode roundrobin | publish
//!   --behavior  source | sink | echo
//!   --count     number of messages to send/receive
//!   --prefix    message prefix; message i is "<prefix><i>"
//!   --delivery  at-most-once | at-least-once   (default at-most-once)
//!   --identity  32 hex chars (16 bytes)         (optional)
//!   --linger    seconds a source waits after sending (default 1.0)
//!
//! A `sink` prints each received message (utf-8) on its own line and exits
//! after `count` messages.

use std::collections::HashMap;
use std::io::Write;
use std::time::Duration;

use aiomsg::{DeliveryGuarantee, Identity, SendMode, Socket};

fn main() -> aiomsg::Result<()> {
    let args = parse_args();
    let rt = tokio::runtime::Runtime::new()?;
    rt.block_on(run(args))
}

fn parse_args() -> HashMap<String, String> {
    let mut map = HashMap::new();
    let mut it = std::env::args().skip(1);
    while let Some(key) = it.next() {
        if let Some(name) = key.strip_prefix("--") {
            if let Some(value) = it.next() {
                map.insert(name.to_string(), value);
            }
        }
    }
    map
}

async fn run(args: HashMap<String, String>) -> aiomsg::Result<()> {
    let role = args.get("role").map(String::as_str).unwrap_or("connect");
    let host = args.get("host").map(String::as_str).unwrap_or("127.0.0.1");
    let port: u16 = args
        .get("port")
        .and_then(|p| p.parse().ok())
        .unwrap_or(25000);
    let behavior = args.get("behavior").map(String::as_str).unwrap_or("sink");
    let count: usize = args.get("count").and_then(|c| c.parse().ok()).unwrap_or(10);
    let prefix = args.get("prefix").map(String::as_str).unwrap_or("m");
    let linger: f64 = args
        .get("linger")
        .and_then(|l| l.parse().ok())
        .unwrap_or(1.0);

    let send_mode = match args.get("send-mode").map(String::as_str) {
        Some("publish") => SendMode::Publish,
        _ => SendMode::RoundRobin,
    };
    let delivery = match args.get("delivery").map(String::as_str) {
        Some("at-least-once") => DeliveryGuarantee::AtLeastOnce,
        _ => DeliveryGuarantee::AtMostOnce,
    };

    let mut builder = Socket::builder()
        .send_mode(send_mode)
        .delivery_guarantee(delivery);
    if let Some(hex) = args.get("identity") {
        builder = builder.identity(parse_identity(hex));
    }
    let sock = builder.build();

    let addr: std::net::SocketAddr = format!("{host}:{port}").parse().expect("valid addr");
    match role {
        "bind" => {
            sock.bind(addr).await?;
        }
        _ => {
            sock.connect(addr).await?;
        }
    }

    match behavior {
        "source" => {
            for i in 0..count {
                sock.send(format!("{prefix}{i}")).await?;
            }
            tokio::time::sleep(Duration::from_secs_f64(linger)).await;
        }
        "echo" => {
            for _ in 0..count {
                if let Some((id, msg)) = sock.recv_identity().await {
                    sock.send_to(id, msg).await?;
                }
            }
            tokio::time::sleep(Duration::from_secs_f64(linger)).await;
        }
        _ => {
            // sink
            let stdout = std::io::stdout();
            let mut received = 0;
            while received < count {
                match sock.recv().await {
                    Some(msg) => {
                        let mut lock = stdout.lock();
                        let _ = lock.write_all(&msg);
                        let _ = lock.write_all(b"\n");
                        let _ = lock.flush();
                        received += 1;
                    }
                    None => break,
                }
            }
        }
    }

    sock.close().await;
    Ok(())
}

fn parse_identity(hex: &str) -> Identity {
    let mut id = [0u8; 16];
    for (i, slot) in id.iter_mut().enumerate() {
        if let Some(byte) = hex.get(i * 2..i * 2 + 2) {
            *slot = u8::from_str_radix(byte, 16).unwrap_or(0);
        }
    }
    id
}
